File: trip-server/src/services/llmGuard/semaphore.ts
export class Semaphore {
  private current = 0

  constructor(private max: number) {}

  tryAcquire(): boolean {
    if (this.current >= this.max) return false
    this.current++
    return true
  }

  release(): void {
    if (this.current > 0) this.current--
  }

  get available(): number {
    return this.max - this.current
  }
}

export class ConcurrencyGuard {
  private global: Semaphore
  private perUser = new Map<string | number, { sem: Semaphore; refs: number }>()
  private cleanupInterval: ReturnType<typeof setInterval>

  constructor(globalMax: number, private perUserMax: number) {
    this.global = new Semaphore(globalMax)
    this.cleanupInterval = setInterval(() => this.cleanup(), 60_000)
    if (this.cleanupInterval.unref) this.cleanupInterval.unref()
  }

  tryAcquire(userId: string | number | null): { success: boolean; release: () => void } {
    let userSlot = false
    let userSem: Semaphore | null = null

    if (userId != null) {
      let entry = this.perUser.get(userId)
      if (!entry) {
        entry = { sem: new Semaphore(this.perUserMax), refs: 0 }
        this.perUser.set(userId, entry)
      }
      userSem = entry.sem
      userSlot = userSem.tryAcquire()
      if (!userSlot) {
        return { success: false, release: () => {} }
      }
      entry.refs++
    }

    const globalSlot = this.global.tryAcquire()
    if (!globalSlot) {
      if (userSlot && userSem && userId != null) {
        userSem.release()
        const entry = this.perUser.get(userId)
        if (entry) entry.refs--
      }
      return { success: false, release: () => {} }
    }

    const releasedUserId = userId
    return {
      success: true,
      release: () => {
        this.global.release()
        if (releasedUserId != null) {
          const entry = this.perUser.get(releasedUserId)
          if (entry) {
            entry.sem.release()
            entry.refs--
          }
        }
      },
    }
  }

  shutdown(): void {
    clearInterval(this.cleanupInterval)
    this.perUser.clear()
  }

  private cleanup(): void {
    for (const [key, entry] of this.perUser) {
      if (entry.refs <= 0) this.perUser.delete(key)
    }
  }
}

File: trip-server-py/src/services/agent/semaphore.py
"""并发守卫模块。

限制并发 LLM 调用数量。
"""

import asyncio
from typing import Callable, Optional


class ConcurrencyGuard:
    """并发守卫。
    
    功能：
    - 全局并发限制
    - 单用户并发限制
    """
    
    def __init__(self, global_max: int = 10, per_user_max: int = 1):
        """初始化并发守卫。
        
        Args:
            global_max: 全局最大并发数
            per_user_max: 单用户最大并发数
        """
        self._global = asyncio.Semaphore(global_max)
        self._per_user: dict[int, asyncio.Semaphore] = {}
        self._per_user_max = per_user_max
        self._lock = asyncio.Lock()
    
    async def try_acquire(self, user_id: int) -> tuple[bool, Optional[Callable]]:
        """尝试获取信号量。
        
        Args:
            user_id: 用户 ID
            
        Returns:
            (成功标志, 释放函数)
        """
        # 全局检查
        if self._global.locked():
            return False, None
        
        # 单用户检查
        async with self._lock:
            if user_id not in self._per_user:
                self._per_user[user_id] = asyncio.Semaphore(self._per_user_max)
            
            user_sem = self._per_user[user_id]
            if user_sem.locked():
                return False, None
        
        # 获取信号量
        await self._global.acquire()
        
        async with self._lock:
            user_sem = self._per_user[user_id]
            await user_sem.acquire()
        
        released = False
        
        async def release() -> None:
            """释放信号量。"""
            nonlocal released
            if released:
                return
            released = True
            self._global.release()
            
            async with self._lock:
                if user_id in self._per_user:
                    self._per_user[user_id].release()
        
        return True, release
    
    async def __aenter__(self):
        """异步上下文管理器入口。"""
        await self._global.acquire()
        return self
    
    async def __aexit__(self, *args):
        """异步上下文管理器出口。"""
        self._global.release()


# 全局单例
concurrency_guard = ConcurrencyGuard()
